src/graph_visualization.cpp

```diff
@@ -0,0 +1,463 @@
+  int flag = 1;
+  while (flag) {
+    flag = 0;
+    for (int i = 0; i < nodes.size(); i++) {
+      for (int j = 0; j < nodes[i].succ.size(); j++) {
+        TreeNode succNode = getTreeNodeById(nodes, nodes[i].succ[j]);
+        if (nodes[i].layer + 1 < succNode.layer) {
+          addDummyNode(nodes, lens_layer, nodes[i], succNode, 1);
+          flag = 1;
+          break;
+        }
+      }
+      if (flag == 1) {
+        break;
+      }
+    }
+  }
+
+  for (int i = 0; i < deletedEdges.size(); i++) {
+    TreeNode firstNode = getTreeNodeById(nodes, deletedEdges[i].first);
+    TreeNode secondNode = getTreeNodeById(nodes, deletedEdges[i].second);
+
+    if (firstNode.layer == secondNode.layer + 1) {
+      for (int j = 0; j < nodes.size(); j++) {
+        if (nodes[j].id == firstNode.id) {
+          nodes[j].succ.push_back(secondNode.id);
+        }
+        if (nodes[j].id == secondNode.id) {
+          nodes[j].pred.push_back(firstNode.id);
+        }
+      }
+    } else {
+      int countDummy = firstNode.layer - (secondNode.layer + 1);
+      addDummyNode(nodes, lens_layer, firstNode, secondNode, -1);
+      for (int j = 1; j < countDummy; j++) {
+        addDummyNode(nodes, lens_layer, nodes.back(), secondNode, -1);
+      }
+    }
+  }
+}
+
+void transformationGraph(std::vector<std::vector<int>> &input, std::vector<TreeNode> &output) {
+  initIdAndSucc(input, output);
+
+  std::vector<std::pair<TreeNode::nodeId, TreeNode::nodeId>> deletedEdges = {};
+  greedyFAS(output, deletedEdges);
+
+  std::vector<int> lens_layer = {};
+  algorithmASAP(output, lens_layer);
+
+  addDummyNodes(output, lens_layer, deletedEdges);
+
+  predReinit(output);
+  degReinit(output);
+}
+
+void printInfo(std::vector<TreeNode> &nodes) {
+  for (int i = 0; i < nodes.size(); i++) {
+    if (nodes[i].isDummy == true) {
+      printf("№%i (dummy)| number %i, layer %i, dp %i, dm %i", nodes[i].id, nodes[i].number, nodes[i].layer, nodes[i].degP, nodes[i].degM);
+    } else {
+      printf("№%i| number %i, layer %i, dp %i, dm %i", nodes[i].id, nodes[i].number, nodes[i].layer, nodes[i].degP, nodes[i].degM);
+    }
+    printf("\nsucc: ");
+    for (int j = 0; j < nodes[i].succ.size(); j++) {
+      printf("%i ", nodes[i].succ[j]);
+    }
+    printf("\npred: ");
+    for (int j = 0; j < nodes[i].pred.size(); j++) {
+      printf("%i ", nodes[i].pred[j]);
+    }
+    printf("\n");
+  }
+}
+
+const std::vector<TreeNode::nodeId> &Net::getSources() {
+  degReinit(nodes);
+  for (int i = 0; i < nodes.size(); i++) {
+    if (nodes[i].degM == 0) {
+      sources.push_back(nodes[i].id);
+    }
+  }
+  return sources;
+}
+
+const std::vector<TreeNode::nodeId> &Net::getSinks() {
+  degReinit(nodes);
+  for (int i = 0; i < nodes.size(); i++) {
+    if (nodes[i].degP == 0) {
+      sinks.push_back(nodes[i].id);
+    }
+  }
+  return sinks;
+}
```

Comment:
Assume `Net` is immutable:
- on the *first* call to `getSources` or `getSinks` compute corresponding node set
- on successive calls return the stored collection
- never modify the collections